src/js/utils/MetronomeUtil.js

```diff
@@ -1,70 +1,79 @@
 /**
- * Parse an item into a Job or JobTree - This method will create sub trees if
- * needed to insert the item at the correct location
- * (based on id/path matching).
- * @param {{
- *          id:string,
- *          items:array<({id:string, items:array}|*)>,
- * }} parent tree to add item to
- * @param {{id:string}} item job to add to parent
- * @param {[type]} jobsAlreadyAdded hash of id and jobs of jobs that
- * have already been added to the parent
+ * Get the parent id (e.g. foo) form the iten id (e.g. foo.bar) by matching
+ * everything but the first bit including the preceding dot "." (e.g. ".bar")
+ *
+ * @param {string} itemId
+ * @return {string} parent id
  */
-function addJob(parent, item, jobsAlreadyAdded) {
-  const { id } = parent;
+function getParentId(itemId) {
+  const [parentId = ""] = itemId.match(/.*?(?=\.?[^.]+\.?$)/);
 
-  const itemId = item.id;
+  return parentId;
+}
+
+/**
+ * Add a job to the respective namespace based on the id property.
+ *
+ * @param {{id:string}} job
+ * @param {Map} jobMap hash map of all jobs that have already
+ * been created
+ * @param {Map} namespaceMap hash map of all namespaces that have already
+ * been created
+ * @return {{id:string}} job
+ */
+function addJob(job, jobMap, namespaceMap) {
+  const { id } = job;
 
-  if (itemId.startsWith(".") || itemId.endsWith(".")) {
+  if (id.startsWith(".") || id.endsWith(".")) {
     throw new Error(
-      `Id (${itemId}) must not start with a leading dot (".") ` +
+      `Id (${id}) must not start with a leading dot (".") ` +
         "and should not end with a dot."
     );
   }
 
-  if (!itemId.startsWith(id)) {
-    throw new Error(`item id (${itemId}) doesn't match tree id (${id})`);
-  }
-
-  // Check if the item (group or job) has already been added
-  if (jobsAlreadyAdded[itemId]) {
-    // handle merge data for job, not for group
-    item = Object.assign(jobsAlreadyAdded[itemId], item);
+  // Check if the item (namespace or job) has already been added and merge with
+  // existing one
+  if (jobMap.has(id)) {
+    job = Object.assign(jobMap.get(id), job);
+    jobMap.set(id, job);
 
-    return;
+    return job;
   }
 
-  // Get the parent id (e.g. group) by matching everything but the item
-  // name including the preceding dot "." (e.g. ".name").
-  const [parentId] = itemId.match(/.*?(?=\.?[^.]+\.?$)/);
+  const parentNamespace = getNamespaceWithId(getParentId(id), namespaceMap);
 
-  if (parentId == null) {
-    return;
-  }
+  // Add job to the parent namespace and jobs map
+  parentNamespace.items.push(job);
+  jobMap.set(id, job);
```

Comment:
Yeah, this util is anything but nice and I would love to actually change how we transform the list of jobs into a tree.